qgis/QGIS_Tasks.md

```diff
@@ -0,0 +1,454 @@
+            # if it was canceled by the user
+            QgsMessageLog.logMessage(
+                message=f"Canceled download task. {len(self.tiles)} tiles downloaded.",
+                level=Qgis.Warning,
+            )
+            return
+        elif not result:
+            # if there was an error
+            QMessageBox.critical(
+                iface.mainWindow(),
+                "Tile download error",
+                f"The following error occurred:\n{self.exception.__class__.__name__}: {self.exception}",
+            )
+            return
+
+        iface.messageBar().pushMessage(
+            "Success",
+            f"Finished downloading {len(self.tiles)} tiles.",
+            level=Qgis.Success,
+        )
+
+        # add the rasters to the map canvas
+        for tile_path in self.tiles:
+            if not tile_path.exists():
+                QgsMessageLog.logMessage(
+                    message=f"Tile at {tile_path.resolve()} was removed.",
+                    level=Qgis.Warning,
+                )
+                continue
+            QgsProject.instance().addMapLayer(
+                QgsRasterLayer(str(tile_path.resolve()), tile_path.stem)
+            )
+
+```
+Here, we want to handle three different cases:
+	1. The user canceled the task
+	2. An exception occured
+	3. The task completed successfully
+
+In the first two cases, we only want to communicate the task outcome to the user. Remember that inside `finished()`, we can access objects from the main thread again, so we can simply push a message to the main interface's message bar. For the third case, we communicate the successful completion of the task to the user and additionally add the downloaded tiles to the map canvas.
+
+### 3.2 Wrap up: user feedback and running the task
+
+Lastly, we want to update the progress bar while the tiles are downloading. Remember that we can't directly access the widget we created earlier in `gui/tile_elevation_dlg.py`, but luckily, `QgsTask` comes with a set of signals we can use to communicate with the user while the background task is running. `progressChanged` is the most important one here: we can emit it by calling `self.setProgress(int)`, where `int` is an integer between 0 - 100, indicating the tasks progress in percent. We connect to the signal in `gui/tile_elevation_dlg.py` right after initializing the task, and to the `taskCompleted` signal to clear the message bar again after the task has run:
+
+```python
+def download_tiles(self) -> None:
+    ...
+    self.task = DownloadTask(grid, out_dir)
+    self.task.progressChanged.connect(self.progress_bar.setValue)
+    self.task.taskCompleted.connect(self.iface.messageBar().clearWidgets)
+    QgsApplication.taskManager().addTask(self.task)
+    self.close()
+```
+
+> Note that it's vital to assign the task not just to a variable, but to a class attribute. Otherwise, the task will be deleted after `download_tiles()` finishes, but before the task itself finishes.
```

Comment:
```suggestion
> **Note** that it's vital to assign the task not just to a variable, but to a class attribute. Otherwise, the task will be deleted after `download_tiles()` finishes, but before the task itself finishes.
```